src/game_api/socket.cpp

```diff
@@ -53,41 +51,132 @@ void dump_network()
     {
-        std::optional<std::string> ret = server->cb(std::string(buf, n));
-        if (ret)
+        sockpp::inet_address src;
+        ssize_t n;
+        while ((n = sock.recv_from(buf, sizeof(buf), &src)) > 0)
         {
-            socket.send_to(ret.value(), src);
+            std::optional<std::string> ret = cb(std::string(buf, n), src.to_string());
+            if (ret.has_value())
+                sock.send_to(ret.value(), src);
         }
+        std::this_thread::sleep_for(std::chrono::microseconds(1));
+    }
+    m_opened = false;
+    sock.shutdown();
+}
+UdpServer::UdpServer(std::string host, in_port_t port)
+    : m_address(host, port)
+{
+    if (m_sock.bind(m_address))
+    {
+        m_address = m_sock.address(); // update port etc.
+        m_opened = true;
+        m_sock.set_non_blocking();
     }
 }
+UdpServer::UdpServer(std::string address)
+    : m_address(address_from_string(address))
+{
+    if (m_sock.bind(m_address))
+    {
+        m_address = m_sock.address(); // update port etc.
+        m_opened = true;
+        m_sock.set_non_blocking();
+    }
+}
+void UdpServer::close()
+{
+    m_opened = false;
+    if (!m_thread.joinable())
+        m_sock.close();
+}
+ssize_t UdpServer::send(std::string message, std::string host, in_port_t port)
+{
+    if (!is_open())
+        return -1;
 
-UdpServer::UdpServer(std::string host_, in_port_t port_, std::function<SocketCb> cb_)
-    : host(host_), port(port_), cb(cb_)
+    return m_sock.send_to(message, sockpp::inet_address(host, port));
+}
+ssize_t UdpServer::send(std::string message, std::string address)
 {
-    sock.bind(sockpp::inet_address(host, port));
-    kill_thr.test_and_set();
-    thr = std::thread(udp_data, std::move(sock), this);
+    if (!is_open())
+        return -1;
+
+    return m_sock.send_to(message, address_from_string(std::move(address)));
 }
-void UdpServer::clear() // TODO: fix and expose: this and the destructor causes deadlock
+sockpp::inet_address UdpServer::address_from_string(std::string address)
 {
-    kill_thr.clear(std::memory_order_release);
-    thr.join();
+    auto colon = address.rfind(':');
+    sockpp::inet_address final_address;
+    if (colon == std::string::npos)
+    {
+        final_address.create(address, 0);
+    }
+    else
+    {
+        std::string_view port_str(address);
+        port_str = port_str.substr(colon + 1);
+        in_port_t port = 0;
+        std::from_chars(port_str.data(), port_str.data() + port_str.size(), port);
+        address.resize(colon);
+        final_address.create(address, port);
+    }
+    return final_address;
 }
-UdpServer::~UdpServer()
+ssize_t UdpServer::read(std::function<ReadFun> fun)
+{
+    if (!is_open())
+        return -1;
+
+    static char buf[32768];
+    sockpp::inet_address src;
+    auto ret = m_sock.recv_from(buf, sizeof(buf), &src);
+
+    if (ret > -1)
+        fun(std::string(buf, static_cast<size_t>(ret)), src.to_string());
```

Comment:
Could also make the callback return string that would be send back, just like `udp_listen`.
It doesn't really matter as you can also just use `send` inside that function